File: legacy/packages/planner/src/free-vars.ts
import type { Expression, Pattern } from "@moon/ast";

export function freeVarsExpr(expr: Expression, bound = new Set<string>()): Set<string> {
  switch (expr.kind) {
    case "Var":
      return bound.has(expr.name) ? new Set() : new Set([expr.name]);
    case "Lit":
      return new Set();
    case "App":
      return union(freeVarsExpr(expr.func, bound), freeVarsExpr(expr.arg, bound));
    case "Infix":
      return union(
        freeVarsExpr(expr.left, bound),
        freeVarsExpr(expr.right, bound),
      );
    case "Prefix":
      return freeVarsExpr(expr.operand, bound);
    case "FieldAccess":
      return freeVarsExpr(expr.object, bound);
    case "Lambda": {
      const local = new Set(bound);
      for (const p of expr.params) local.add(p);
      if ("statements" in expr.body) {
        return freeVarsDoBlock(expr.body, local);
      }
      return freeVarsExpr(expr.body, local);
    }
    case "If":
      return union(
        freeVarsExpr(expr.condition, bound),
        freeVarsExpr(expr.thenBranch, bound),
        freeVarsExpr(expr.elseBranch, bound),
      );
    case "Record":
      return union(...expr.fields.map((f) => freeVarsExpr(f.value, bound)));
    case "List":
      return union(...expr.elements.map((e) => freeVarsExpr(e, bound)));
    case "Tuple":
      return union(...expr.elements.map((e) => freeVarsExpr(e, bound)));
    case "Paren":
      return freeVarsExpr(expr.expr, bound);
    case "Do":
      return freeVarsDoBlock(expr.block, bound);
    case "Agent":
    case "Model":
      return new Set();
  }
}

export function freeVarsDoBlock(
  block: { statements: import("@moon/ast").DoStatement[] },
  bound = new Set<string>(),
): Set<string> {
  const result = new Set<string>();
  const local = new Set(bound);

  for (const stmt of block.statements) {
    if (stmt.kind === "Let" || stmt.kind === "Bind") {
      for (const v of freeVarsExpr(stmt.expr, local)) result.add(v);
      bindPatternVars(stmt.pattern, local);
    } else if (stmt.kind === "Storm") {
      for (const v of freeVarsExpr(stmt.input, local)) result.add(v);
      for (const item of stmt.config) {
        for (const v of freeVarsExpr(item.value, local)) result.add(v);
      }
      bindPatternVars(stmt.pattern, local);
    } else {
      for (const v of freeVarsExpr(stmt.expr, local)) result.add(v);
    }
  }

  return result;
}

function bindPatternVars(pat: Pattern, bound: Set<string>): void {
  switch (pat.kind) {
    case "PVar":
      bound.add(pat.name);
      break;
    case "PCon":
      for (const arg of pat.args) bindPatternVars(arg, bound);
      break;
    case "PTuple":
      for (const el of pat.elements) bindPatternVars(el, bound);
      break;
    case "PList":
      for (const el of pat.elements) bindPatternVars(el, bound);
      break;
    default:
      break;
  }
}

function union(...sets: Set<string>[]): Set<string> {
  const result = new Set<string>();
  for (const s of sets) {
    for (const v of s) result.add(v);
  }
  return result;
}

File: Sources/MoonPlanner/FreeVars.swift
import MoonAST

func freeVarsExpr(_ expr: Expression, bound: Set<String> = []) -> Set<String> {
    switch expr {
    case .varRef(let name, _):
        return bound.contains(name) ? [] : [name]
    case .lit:
        return []
    case .app(let fn, let arg, _):
        return freeVarsExpr(fn, bound: bound).union(freeVarsExpr(arg, bound: bound))
    case .infix(_, let left, let right, _):
        return freeVarsExpr(left, bound: bound).union(freeVarsExpr(right, bound: bound))
    case .prefix(_, let operand, _):
        return freeVarsExpr(operand, bound: bound)
    case .fieldAccess(let object, _, _):
        return freeVarsExpr(object, bound: bound)
    case .lambda(let params, let body, _):
        var local = bound
        for p in params { local.insert(p) }
        switch body {
        case .expression(let e):
            return freeVarsExpr(e, bound: local)
        case .doBlock(let block):
            return freeVarsDoBlock(block, bound: local)
        }
    case .ifExpr(let condition, let thenBranch, let elseBranch, _):
        return freeVarsExpr(condition, bound: bound)
            .union(freeVarsExpr(thenBranch, bound: bound))
            .union(freeVarsExpr(elseBranch, bound: bound))
    case .record(_, let fields, _):
        return fields.reduce(into: Set<String>()) { acc, field in
            acc.formUnion(freeVarsExpr(field.value, bound: bound))
        }
    case .list(let elements, _):
        return elements.reduce(into: Set<String>()) { acc, el in
            acc.formUnion(freeVarsExpr(el, bound: bound))
        }
    case .tuple(let elements, _):
        return elements.reduce(into: Set<String>()) { acc, el in
            acc.formUnion(freeVarsExpr(el, bound: bound))
        }
    case .paren(let inner, _):
        return freeVarsExpr(inner, bound: bound)
    case .doExpr(let block, _):
        return freeVarsDoBlock(block, bound: bound)
    case .agent, .model:
        return []
    }
}

func freeVarsDoBlock(_ block: DoBlock, bound: Set<String> = []) -> Set<String> {
    var result = Set<String>()
    var local = bound

    for stmt in block.statements {
        switch stmt {
        case .letBind(let pattern, let expr, _):
            result.formUnion(freeVarsExpr(expr, bound: local))
            bindPatternVars(pattern, bound: &local)
        case .bind(let pattern, let expr, _, _):
            result.formUnion(freeVarsExpr(expr, bound: local))
            bindPatternVars(pattern, bound: &local)
        case .storm(let pattern, let input, let config, _):
            result.formUnion(freeVarsExpr(input, bound: local))
            for item in config {
                result.formUnion(freeVarsExpr(item.value, bound: local))
            }
            bindPatternVars(pattern, bound: &local)
        case .action(let expr, let config, _):
            result.formUnion(freeVarsExpr(expr, bound: local))
            for item in config {
                result.formUnion(freeVarsExpr(item.value, bound: local))
            }
        }
    }

    return result
}

private func bindPatternVars(_ pat: Pattern, bound: inout Set<String>) {
    switch pat {
    case .pVar(let name, _):
        bound.insert(name)
    case .pCon(_, let args, _):
        for arg in args { bindPatternVars(arg, bound: &bound) }
    case .pTuple(let elements, _):
        for el in elements { bindPatternVars(el, bound: &bound) }
    case .pList(let elements, _):
        for el in elements { bindPatternVars(el, bound: &bound) }
    case .pWildcard, .pLit:
        break
    }
}